[ECS] Implement JsonFileJournal and use it as SQLiteJournal replacement

mkilmanas opened the pull request:
Fixes #3183

mkilmanas commented:
By the way - this implementation is rather slow and disk-intensive as it reads and rewrites the journal file on each modification (i.e. each call to `write()` or `clean()`).

I could probably refactor this to do read/write only once and keep the file opened/locked for the whole duration of the process and keep the data in memory at that time. But I'm just not sure which one is the higher priority: a) overall speed vs b) journal file availability to any other processes while ECS is running. Current implementation favours the latter

TomasVotruba commented:
Thanks for the PR :+1: 

I've allowed the CI to give the feedback. It must pass before review

mkilmanas commented:
Still working on the PHPStan issues. But there are some that I don't quite understand:
- `"empty($dependencies[\Nette\Caching\Cache::TAGS])" is forbidden to use` - why? a) I've copied this one from Nette; b) what's wrong with checking if array key is empty (either undefined or contains an empty array)? What am I supposed to use instead then?

mkilmanas commented:
@TomasVotruba I have fixed PHPStan issues and rebased on top of latest main branch - hopefully that solves the PHPUnit problems as they were happening way outside of ECS scope and I can't seem to run the full suite locally (phpunit dies after a while with "include() failed - too many open files").

Please allow the checks to run once again.

samsonasik commented:
@mkilmanas applied allow check

mkilmanas commented:
Thanks. However I just realized that now it has logical issues, so there will be another update soon

mkilmanas commented:
Hopefully this is it.

Though I would like someone to dig through the code to verify it. Ironically, while trying to minimize the cognitive complexity of one class it has now grown into quite a few classes, and I'm not so sure if that makes it easier to understand what's going on.

TomasVotruba commented:
I'm wondering, if there is some better and simpler solution. Possibly without nette/cache? All we need to save hash for config and one category.

TomasVotruba commented:
I'm wondering, if there is some better and simpler solution. Possibly without nette/cache? All we need to save hash for config and one category.

mkilmanas commented:
> I'm wondering, if there is some better and simpler solution. Possibly without nette/cache? All we need to save hash for config and one category.

So you're thinking a file like this would be enough of a "cache"?
```json
{
    "src/Foo.php": "db940ebadadb9d64a8c02718749fd2c5",
    "src/Controller/BarController.php": "66f8401f426b3cedc450b32e9edc50eb",
    ...
}
```

What are the tags currently used for anyway? Aren't you interested only if a) the file ins question; b) the config file has changed since last run?

mkilmanas commented:
@TomasVotruba - any views/suggestions on how we could proceed from here?

I tend to agree that this solution looks like a bit of an overkill, but I would like to know the expectations for acceptable approach and a deeper understanding on the use-case of the caching here.

TomasVotruba commented:
Hi, I'm sorry I missed the notification here. I'm looking into this now...

TomasVotruba commented:
> So you're thinking a file like this would be enough of a "cache"?

Yes, I think that's what is used here.

The tags are used only to invalidate all file cache if config has changed.

TomasVotruba commented:
Thank you! I'll see what I can do about symplifing it and possibly remove nette/caching and use your bare implementatoin.

TomasVotruba commented:
I'm trying to build on your work and make it create our custom cache classes - with 0-dependency on external frameworks.: https://github.com/symplify/symplify/pull/3284/files

mkilmanas commented:
Great - whichever way makes it work is great : 👍 😄
